rts/lra/lra-test/lra-test-arquillian-extension/src/main/java/io/narayana/lra/arquillian/spi/NarayanaLRARecovery.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2020, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package io.narayana.lra.arquillian.spi;
+
+import io.narayana.lra.LRAConstants;
+import io.narayana.lra.logging.LRALogger;
+
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+
+public class NarayanaLRARecovery {
+
+    public void waitForCallbacks(URI lraId) {
+        // no action is needed, tck callbacks calls are sufficiently fast for TCK
+    }
+
+    // TODO remove once this class implements LRARecoveryService
+    public void waitForRecovery(URI lraId) {
+        int counter = 0;
+
+        do {
+            LRALogger.logger.info("Recovery attempt #" + ++counter);
+        } while (!waitForEndPhaseReplay(lraId));
+        LRALogger.logger.info("LRA " + lraId + "has finished the recovery " + counter);
+    }
+
+    public boolean waitForEndPhaseReplay(URI lraId) {
+        String host = lraId.getHost();
+        int port = lraId.getPort();
+        if (!recoverLRAs(host, port, lraId)) {
+            // first recovery scan probably collided with periodic recovevery which started
+            // before the test execution so try once more
+            return recoverLRAs(host, port, lraId);
```

Comment:
@xstefank @mmusgrov perhaps we could turn off automated recovery for this test if we want more predictability

Reply:
@tomjenkinson AfterLRA can be replayed several times successfully. The test is just testing that it will be eventually replayed, but the specification allows it to be replayed even 4 or 5 times (unbounded). So I think the test is correct.

Reply:
Sure, but under test conditions in Narayana we can control this to reduce the number of retries. If we turned off background recovery and triggered the recovery manually you would only need one call to recoverLRAs(host, port, lraId).

However, given the recovery manager is (I assume) running in a service similar to in WildFly then actually this is not necessarily possible. What might be possible is to delay the background recovery thread from starting for some arbitrarily long period using https://github.com/jbosstm/narayana/blob/1e17caa16b5efb9b16cdb2d8489a1b93722ebdd5/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/common/RecoveryEnvironmentBean.java#L47

Reply:
@tomjenkinson thanks, I didn't know about this. We can use I think. But let's leave that to some other PR/issue, please.

Reply:
Yeah, seems reasonable as the test is valid either way I think as we are getting the extra notifications.

Reply:
We already [tweak that config option](https://github.com/jbosstm/narayana/blob/master/rts/lra/lra-test/pom.xml#L118). You could change it now on the LRA branch or just wait for the merge into master. I don't think there is any particular urgency.